### PR TITLE
Workflow improvements: helpers pkg with suite generator base, and automatic makefile targets

### PR DESCRIPTION
**Pull request by protolambda**
To speed up and ease the creation of new test generators, I implemented a base package which makes it really easy to create new generators with. The format conforms to my proposed suite format in #29 

I also updated the makefile, as most people hate writing makefiles, when they could be spending time on new test-generators/research.

The idea is basically that top-level folder names define the test-types. In these folders, there's a bunch of yaml files, which we call "suites". Each suite has a name, summary, constants-config, fork, and a test-cases list.

The goal of a generator is to a set of suites that cover different facets of the spec-topic (test type) being tested.

## How generators look like

This PR included just the helper package at first, but since it's more about workflow, I decided to make it streamlined: people should not have to touch the makefile, just write a test-generator. And with the helpers package, this should be super easy.

Check the new README for an example on how to write a new test generator.

## WIP
Work in progress. What we need to get this running:
- accept issue #29 as standard (it doesn't change much of existing code except for preventing config duplication)

Note that this does not break any existing generators. Although I do hope existing generators move to adopt this in the future. SSZ is the easiest to port, see below. Note that the existing generators are not exactly conformaing to my proposed suite header data, but still work, and output still in the same way.

## Credits

Much of the generator code has been adapted from the previous SSZ test generator. The object yield pattern is very nice, and command-line arguments are great to have too.

### CODE REVIEW

**Comment by protolambda** [comment]
Streamlined it some more. Everyone can easily write and work on new test-generators now :tada:

**Comment by mratsim** [comment]
Awesome work :)

File: test_libs/gen_helpers/gen_base/gen_suite.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+from eth_utils import (
+    to_dict,
+)
+
+from gen_base.gen_typing import TestSuiteTestCases
+
+
+@to_dict
+def render_suite(*, title: str, summary: str, fork: str, config: str, test_cases: TestSuiteTestCases):
```
Comment (jannikluhn) [inline test_libs/gen_helpers/gen_base/gen_suite.py]:
```suggestion
def render_suite(*, title: str, summary: str, fork: str, config: str, test_cases: Iterable[TestCase]):
```

should suffice and allows users to pass a generator or a list if they want to for some reason. Alternatively, define `TestSuiteTestCases = Iterable[TestCase]` in `gen_typing.py`, but I don't see why.

**Review by jannikluhn — APPROVED** [review]

File: test_libs/gen_helpers/gen_base/gen_suite.py (hunk 1 above)
Comment (protolambda) [reply test_libs/gen_helpers/gen_base/gen_suite.py]: Good suggestion. Went a little overboard with typing there. Also fixed a bit of formatting. See 
25dbfee

File: test_libs/gen_helpers/gen_base/gen_suite.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+
+@to_dict
+def render_suite(*, title: str, summary: str, fork: str, config: str, test_cases: Iterable[TestCase]):
+    yield "title", title
+    if summary is not None:
+        yield "summary", summary
+    yield "fork", fork
+    yield "config", config
+    yield "test_cases", test_cases
```
Comment (jannikluhn) [inline test_libs/gen_helpers/gen_base/gen_suite.py]: Oops, just noticed, if we allow test cases to be iterables this should probably become something like
```suggestion
    yield "test_cases", tuple(test_cases)
```

**Comment by protolambda** [comment]
Implemented in https://github.com/ethereum/eth2.0-specs/pull/851

**Pull request closed without merging**